### PR TITLE
[#31203] The one way and the one container for store the script options

### PR DESCRIPTION
**Pull request by Fedik**
Wiki: [Adding the options to your JavaScript code](https://docs.joomla.org/J3.x:Adding_JavaScript_and_CSS_to_the_page#Adding_the_options_to_your_JavaScript_code)

Example add the script option:

``` php
$options = array(
  'showAlert' => false,
  'consolelog' => true,
  'text' => 'Hi! I am "Text"!',
  'someArray' => array('one', 'two', 'and so on')
);
$doc->addScriptOptions('plg_examplejs', $options);
```

Access to the options in frontend:

``` javascript
Joomla.optionsStorage.plg_examplejs;
```

**Links:**
- old pull #1319
- [[#31203] The one way and the one container for store the script options](http://joomlacode.org/gf/project/joomla/tracker/?action=TrackerItemEdit&tracker_item_id=31203) - feature request

### CODE REVIEW

**Comment by yvesh** [comment]
Works without problems and offers a nice way to store javascript options :+1: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/3072">issues.joomla.org/joomla-cms/3072</a>.</sub>

**Comment by anibalsanchez** [comment]
It is similar to #6006 but #6006 is more compact<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/3072">issues.joomla.org/joomla-cms/3072</a>.</sub>

**Comment by Kubik-Rubik** [comment]
@Fedik Since we don't have a reaction for so long, is this still a valid feature request?

@dgt41 Could you please also review here? Thanks!

**Comment by Fedik** [comment]
@Kubik-Rubik yes, this is still a very very very very valid feature request :wink: 
at least I hope so :smile:

**Comment by roland-d** [comment]
@Fedik I am not understanding the use of this PR. Can you explain it in more detail?

File: libraries/joomla/document/document.php [hunk 1]
```diff
@@ -525,6 +532,56 @@ public function addScriptDeclaration($content, $type = 'text/javascript')
...
+	 *
+	 * @since   5.0
+	 */
+	public function addScriptOptions($key, $options, $merge = true)
+	{
+		if (empty($this->scriptOptions[$key]))
+		{
+			$this->scriptOptions[$key] = array();
+		}
+		if ($merge && is_array($options))
```
Comment (roland-d) [inline libraries/joomla/document/document.php]: Requires a blank line before the if statement according to codestyle.

File: libraries/joomla/document/html/renderer/head.php [hunk 2]
```diff
@@ -194,6 +194,35 @@ public function fetchHead($document)
 			$buffer .= '></script>' . $lnEnd;
 		}
 
+		// Generate scripts options
+		$scriptOptions = $document->getScriptOptions();
+		if (!empty($scriptOptions))
```
Comment (roland-d) [inline libraries/joomla/document/html/renderer/head.php]: Requires a blank line before the if statement according to codestyle

**Comment by Fedik** [comment]
@roland-d for now if you need to add some JavaScript variable for one of your script you do:
```php
$doc->addScriptDeclaration("
var foo = 'blabla';
var bar = ['more', 'blabla'];
");
```
Such approach have couple problem:
- you cannot easily modify it at runtime , only by parsing `$doc->_scripts` with some regex;
- this will add these variables in Global scope that can overlap the existing variable from another extension.

With my suggestion you can do:
```php
$doc->addScriptOptions('myextension_name', array(
  'foo' =>  'blabla',
  'bar' => ['more', 'blabla']
));
```
and in javascript access to them:
```javascript
console.log(Joomla.optionsStorage.myextension_name.foo);
console.log(Joomla.optionsStorage.myextension_name.bar);
```
As you can see all options under `Joomla.optionsStorage.myextension_name` instead of global scope.

Also you can modify your options at PHP runtime easily, eg. from plugin or where you need to do it.
```php
$options = $doc->getScriptOptions('myextension_name');
$options['foo'] = 'ue ue';
$doc->addScriptOptions('myextension_name', $options);

// or more easy:
$doc->addScriptOptions('myextension_name', array('foo' => 'ue ue')); 
// it will be merged, and old 'foo' changed
```

Initially I made it for  #1260, #6357, as with such approach we can move 99% of inline javascript to the files more easy, and rule them just by using the appropriate options in `Joomla.optionsStorage`

**Comment by dgt41** [comment]
@Fedik I was speaking with some PLT members about the possibility to use more the data attributes to pass strings or PHP values. This way we drop the inline script just to pass some simple (or more complex data) from PHP to Javascript. Although a way to do it with inline script MUST also be available. So my idea is to drop JText::script() and replace it with this function you introduce here, and make sure that in most places (at least in the core) we just use data attributes. But that is more a 4.0 thing I guess...

**Comment by peterpeter** [comment]
@dgt41 And what means your comment to someone who wants testing patches? No further tests on this one?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/3072">issues.joomla.org/joomla-cms/3072</a>.</sub>

**Comment by roland-d** [comment]
@Fedik Can this be used next to what we have now? In that case we may be able to work on this for 3.6 as there seems general consensus this is the way forward.

**Comment by Fedik** [comment]
@roland-d yes, it still valid,
I just need to check the conflicts, someday :smiley:

**Comment by roland-d** [comment]
@Fedik I had another go at this PR and it's great. I now fully understand what it does and see great potential for it :) Thanks for the follow-up even though it has taken 2 years. Let's get it done for 3.6.

**Comment by roland-d** [comment]
I have tested this item :white_check_mark: successfully on 852311a45e37d0a109897d12e58a8db7f54d61b6<br /><br />I have added the example array to the view.html.php file of com_content on the frontend. After that loaded the page and I was able to access the JavaScript options using the console editor.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/3072">issues.joomla.org/joomla-cms/3072</a>.</sub>

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on 852311a45e37d0a109897d12e58a8db7f54d61b6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/3072">issues.joomla.org/joomla-cms/3072</a>.</sub>

**Comment by roland-d** [comment]
We have 2 successful tests, setting to RTC.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/3072">issues.joomla.org/joomla-cms/3072</a>.</sub>

**Comment by brianteeman** [comment]
Setting to needs review. As the pretty print requires a php version higher than the joomla minimun<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/3072">issues.joomla.org/joomla-cms/3072</a>.</sub>

**Comment by Fedik** [comment]
@brianteeman `JSON_PRETTY_PRINT` in use when it available :wink: 
```php
$pretyPrint  = (JDEBUG && defined('JSON_PRETTY_PRINT') ? JSON_PRETTY_PRINT : false);
```

**Comment by brianteeman** [comment]
Thanks for confirming that - all good - setting back to RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/3072">issues.joomla.org/joomla-cms/3072</a>.</sub>

**Comment by Kubik-Rubik** [comment]
Thank you @Fedik and all testers!

**Pull request merged**